compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirTypeParameterInQualifiedAccessChecker.kt

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2010-2021 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.fir.analysis.checkers.expression
+
+import org.jetbrains.kotlin.fir.analysis.checkers.context.CheckerContext
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.expressions.FirCallableReferenceAccess
+import org.jetbrains.kotlin.fir.expressions.FirQualifiedAccessExpression
+import org.jetbrains.kotlin.fir.resolve.diagnostics.ConeTypeParameterInQualifiedAccess
+import org.jetbrains.kotlin.fir.types.FirErrorTypeRef
+import org.jetbrains.kotlin.fir.types.FirTypeRef
+
+object FirTypeParameterInQualifiedAccessChecker: FirQualifiedAccessChecker() {
+    override fun check(expression: FirQualifiedAccessExpression, context: CheckerContext, reporter: DiagnosticReporter) {
+        checkExplicitReceiver(expression, context, reporter)
+        checkExpressionItself(expression, context, reporter)
+    }
+
+    private fun checkExpressionItself(
+        expression: FirQualifiedAccessExpression,
+        context: CheckerContext,
+        reporter: DiagnosticReporter
+    ) {
+        // Make sure the current qualified access is not part of another qualified access.
+        // E.g., for `T::toString`, which is a callable reference (a subtype of qualified access), type parameter T is checked once as an
+        // explicit receiver (or LHS). When we visit `T` (as a qualified access), we should not regard it as an expression here.
+        if (context.qualifiedAccesses.size > 1) return
+
+        val diagnostic = expression.typeRef.coneTypeParameterInQualifiedAccess ?: return
+        val source = expression.source ?: return
+        reporter.reportOn(source, FirErrors.TYPE_PARAMETER_IS_NOT_AN_EXPRESSION, diagnostic.symbol, context)
+    }
+
+    private fun checkExplicitReceiver(
```

Comment:
Looks like this is because `org.jetbrains.kotlin.fir.analysis.collectors.AbstractDiagnosticCollector` does not override `visitErrorTypeRef`, which, by default, visits resolved type ref and hence skips reporting errors.

It looks like a bug to me and should be fixed with an overridden visitErrorTypeRef. For example

```
override fun visitErrorTypeRef(errorTypeRef: FirErrorTypeRef, data: Nothing?) {
    errorTypeRef.delegatedTypeRef?.accept(this, data)
}
```

Reply:
Thanks for the question. I think my answers are, everything is intentional.

> Should this check be done in org.jetbrains.kotlin.fir.analysis.diagnostics.ConeDiagnosticToFirDiagnosticKt#toFirDiagnostic instead?

It's the place a cone diagnostic can be mapped to a FIR diagnostic without any context, while this one needs to be reported differently, depending on the context, like whether the containing expression is a callable reference, whether this type parameter is a standalone expression, etc.

Re: `AbstractDiagnosticCollector#visitErrorTypeRef`

I'm afraid it will generate lots of redundant/repeated diagnostics.

The reason this diagnostic is not visited is, we intentionally change its source as _fake_ source (as previously mentioned in the code comment).

Reply:
I see. Thanks for the clarification!